bnb_yaz/new_bnb_downloader.py

```diff
@@ -0,0 +1,344 @@
+
+    Logging(progress_info, 1)
+
+    return [ids_info[2], ids_info[3], ids_info[4]]
+
+
+# Convert PDF to text and extract BNB IDs
+def ExtractBNBIDsFromPDF():
+    tmp_pdf_text_file = working_directory + "temp_pdf_text_" + pdf_file_name + ".txt"
+    os.system("pdftotext " + working_directory + pdf_file_name + " " + tmp_pdf_text_file)
+    bnb_ids = []
+    try_to_get = False 
+    with open(tmp_pdf_text_file, "r") as pdf_text_file:
+        for line in pdf_text_file:
+            line = line.strip()
+            if line[:5] == "DDC 2":
+                try_to_get = True 
+                continue
+
+            if line[:10] == "BNB NUMBER" and try_to_get:
+                bnb_ids.append(line[11:len(line)])
+                try_to_get = False
+
+    os.remove(tmp_pdf_text_file)
+    return bnb_ids
+
+
+def UploadToBSZFTPServer(remote_folder_path: str, marc_filename: str):
+    remote_file_name_tmp: str = marc_filename + ".tmp"
+
+    ftp = bsz_util.GetFTPConnection()
+    ftp.changeDirectory(remote_folder_path)
+    ftp.uploadFile(marc_filename, remote_file_name_tmp)
+    ftp.renameFile(remote_file_name_tmp, marc_filename)
+
+# Main function to orchestrate the workflow
+def Main():
+    if(len(sys.argv) != 3):
+        print("usage: " + sys.argv[0] + " <email> <working_directory = "
+        )
+        sys.exit(-1)
+
+    # It is necessary to declare global variables to modify them inside the function
+    global working_directory, pdf_file_name, marc_output_file, email_recipient, db_connection, bnb_yaz_client
+
+    email_recipient = sys.argv[1]
+    working_directory = sys.argv[2]
+    
+    # Check if the working directory exists
+    Logging("Checking working directory: " + working_directory, 1)
+    CheckWorkingDirectory()
+    Logging("Working directory is valid.", 1)
+
+    # Check if the configuration file exists and get the input PDF file name
+    Logging("Getting PDF file name from configuration file.", 1)
+    pdf_file_name = GetPDFFileName()
+    Logging("Input PDF file name: " + pdf_file_name, 1)
+
+    # Prepare other file names
+    Logging("Preparing other file names.", 1)
+    marc_output_file = "BNB_records_" + datetime.datetime.now().strftime("%Y%m%d_%H%M%S") + ".mrc"
+    Logging("MARC output file name: " + marc_output_file, 1)
+    
+    # Initialize the database
+    Logging("Initializing the database.", 1)
+    db_connection = InitializeDatabase()
+    Logging("Database initialized.", 1)
+
+    # Connect to BNB server
+    Logging("Connecting to BNB server.", 1)
+    bnb_yaz_client = ConnectToBNBServer()
+    Logging("Connected to BNB server.", 1)
+
+    # Retrying previously not found BNB IDs
+    Logging("Retrying previously not found BNB IDs.", 1)
+    retrying_download = RetryingPreviousNotFoundBNBIDs()
+    Logging("Retried previously not found BNB IDs.", 1)
+
+    # Extract BNB IDs from the input PDF file
+    Logging("Extracting BNB IDs from the input PDF file.", 1)
+    new_bnb_ids = ExtractBNBIDsFromPDF()
```

Comment:
Also as mentioned before, i would suggest you seperate multiple parts of the process
1) Process the PDF files and write all found IDs into the SQLite database (and move them to the "loaded" folder)
2) Query all non-harvested IDs from the SQLite database and try to harvest them (no matter which PDF file they come from), this way you can process the old ones and the new ones at the same time